wicked.7.txt
wicked(7)
=========

NAME
----

wicked - a lua library for dynamic widgets in awesome

SYNOPSIS
--------

None

DESCRIPTION
-----------

+wicked+ is a library, written in lua, for easy creation and management of dynamic awesome statusbar widgets, from the rc.lua configuration file.

USAGE
-----

To use wicked, put *require("wicked")* at the top of your rc.lua, from then on, create your widgets as normal and call *wicked.register()* to register them to wicked.

'wicked.register(*widget*, *widget_type*, *format*, *update_interval*, *data_field*, *padding*)'


'widget'

Use the widget created by widget for this argument.

'widget_type'

Enter one of the possible widget type functions or a customised one here, see the _WIDGET TYPES_ section.

'format'

Format can be either a string argument where $1,$2,$3,etc. will be replaced by their respective values, or a function called with the arguments *(widget, arguments)*, arguments being a table of the values otherwise inserted into the string, if it is a function, its return value will be used to populate the widget. Note that some widget types might not need or use this parameter.

'update_interval'

The amount of seconds it waits between updating this widget, defaults to 1.

'data_field'

If you want to feed the data into a graph or progressbar, put the bar or plot name you want to use in here.

'padding'

To prevent widgets jumping about your statusbar all the time, you can specify a padding value here, the number given here represents the minimum amount of digits that the widget's output has to contain (ie specifying 2 padds values under 10 with one zero, specifying 3 padds values under 100 with one zero and values under 10 with two zeros, etc)
If the widget type supports it, you can specify a table with different numbers, the first number being the padding for the first value the type would return, the second number for the second value, etc.


OTHER FUNCTIONS
---------------

The following functions are also provided by wicked.

'wicked.unregister(*widget*, *keep*)'

Unregister all updates associated with a widget. Keep is optional (defaults to
false), and if true, the widget will only be suspended, ready to be
reactivated by the activate function later.

'wicked.suspend()'

Suspend all widget updates, useful if you are running in low-power mode and
want to save battery for a while.

'wicked.activate(widget)'

Restarts all widgets that were suspended either from wicked.suspend(), or by
calling wicked.unregister() with keep=true. Widget is optional, if given, only
that widget will be activated, defaults to all widgets.

'wicked.regregister(reg)'

wicked.register() and wicked.unregister() both return a reg object, this
contains information about the widget and the updates that should be sent to
it, you can pass this reg object into regregister after unregistering it, and
it will be reregistered (or activated if it was only suspended).

'wicked.enable_caching(widget_type)'

Enable caching of the values returned by a custom widget_type.
Note that all widget_type functions included in wicked already have caching 
enabled by default.


WIDGET TYPES
------------

Widget types consist of functions that take the "format" argument given to wicked.register as the first argument, the "padding" argument as the second, and return a table of values to insert in the format string, creators of custom widget types should note that the format argument may be a function as well as a string.

For backwards compatibility purposes, if a string X is given as widget_type, it will use \'wicked.widgets.X' as the widget type.

**wicked.widgets.date**

The date widget type uses the output of the \`os.date` command to populate the widget, if format is given, it will be passed as the FORMAT argument to \`os.date`.


**wicked.widgets.cpu**

The cpu widget type outputs the current cpu usage in percentages, calculated over the duration of it's update_interval. Use $1 inside the format string to display the total usage percentage, $2 to show it for your first cpu core only, $3 for the second core, etc.


**wicked.widgets.mem**

The memory widget type outputs four things, $1 will be replaced with the current memory usage in percentages, $2 will be replaced with the total used memory in Mb, $3 with the total amount memory available to the system, and $4 with the amount of free memory.


**wicked.widgets.swap**

The swap widget type outputs the same values as the memory widget type, about the amount of swap in use/available.


**wicked.widgets.mpd**

The mpd widget type outputs the song currently playing on the music player daemon, note that the \`mpc` application needs to be installed and configured in order to use this.


**wicked.widgets.mocp**

The mocp widget type outputs the song currently playing on the mocp server. Use $1 inside the format string to display the current song. A maxlength can be given to the wicked.register() function which will cut the songname in two when exeeded and displays the first and last part. A typical functioncall looks like this: wicked.register(mocpwidget, \'mocp', \'MOC: $1', 1, nil, 50). Here 50 is the max amount of characters displayed of $1.


**wicked.widgets.fs**

The fs widget type can display size and usage information about currently mounted filesystems, use ${mountpoint datatype} inside the format string to display that data, e.g. ${/ avail} displays the free space on the partition mounted on /. Available datatypes are \'used', \'size', \'usep' (Usage in %) and \'avail'.


**wicked.widgets.net**

The net widget type displays usage information about your networks, in the format string, ${interface type} will be replaced with with that value, possible types are: rx, tx, up, down. You can suffix the type with _b, _kb, _mb or _gb, to display the information in those units. 

**wicked.widgets.uptime**

The uptime widget type displays information about how long your machine has been on (or "up"). Use $1 for the total number of seconds up, $2 for the number of days, $3 for the number of hours, $4 for the number of minutes and $5 for the number of seconds.


CUSTOM TYPES
------------

You can create your own custom widgets easily, an example of a widget that returns the output of a specific command is:

--------------------------------------
mywidget = widget({
    type = 'textbox',
    name = 'mywidget'
})

function run_script()
   local filedescriptor = io.popen('my_nifty_script.py')
   local value = filedescriptor:read()
   filedescriptor:close()

   return {value}
end

-- Runs 'my_nifty_script.py' every 10 seconds and puts its output into the widget
wicked.register(mywidget, run_script, "$1", 10)
--------------------------------------

As you can see, the type function returns a table of values to insert into the format string, the first value in the table will be used in place of $1, the second in place of $2, etc.

If you want to register the same function multiple times, it is a good idea to enable caching so that your function is only called once in each interval.
To do this call the 'enable_caching' function like so:

--------------------------------------
wicked.enable_caching(run_script)
--------------------------------------

FORMAT FUNCTIONS
----------------

You can use a function instead of a string as the format parameter, with this, you are able to check the value returned by wicked and change the output accordingly.
The format function will get the widget as its first argument, and a table with the values otherwise inserted into the format string as its second argument, and should return the text to be used for the widget.

An example that uses a format function in order to hide the mpd widget when there is no song playing:

--------------------------------------
mpdwidget = widget({
    type = 'textbox',
    name = 'mpdwidget'
})

wicked.register(mpdwidget, wicked.widgets.mpd, 
    function (widget, args)
        if args[1]:find("volume:") == nil then
            return ' <span color="white">Now Playing:</span> '..args[1]
        else
            return ''
        end
    end)
--------------------------------------


USAGE EXAMPLES
--------------

Put these snippets into your rc.lua (usually located at ~/.config/awesome/rc.lua) to see what they do, don't forget to add the widgets to your statusbar.

**A simple \'Now Playing' widget**

--------------------------------------
mpdwidget = widget({
    type = 'textbox',
    name = 'mpdwidget'
})

wicked.register(mpdwidget, wicked.widgets.mpd,
    ' <span color="white">Now Playing:</span> $1')
--------------------------------------

**A memory usage monitor, with padding**

--------------------------------------
memwidget = widget({
    type = 'textbox',
    name = 'memwidget'
})

wicked.register(memwidget, wicked.widgets.mem,
    ' <span color="white">Memory:</span> $1 ($2Mb/$3Mb)',
    {2, 4, 4})
---------------------------------------

**A filesystem usage indicator**

---------------------------------------
fswidget = widget({
    type = 'textbox',
    name = 'fswidget'
})

wicked.register(fswidget, wicked.widgets.fs,
    ' <span color="white">FS:</span> ${/ used}/${/ size} (${/ usep} used)', 120)
---------------------------------------

**A network interface monitor (for eth0)**

---------------------------------------
netwidget = widget({
    type = 'textbox',
    name = 'netwidget'
})

wicked.register(netwidget, wicked.widgets.net, 
    ' <span color="white">NET</span>: ${eth0 down} / ${eth0 up} [ ${eth0 rx} // ${eth0 tx} ]')
---------------------------------------

SEE ALSO
--------
awesome(1) awesomerc(5) 

AUTHORS
-------
Lucas de Vries <lucas@glacicle.com>


wicked.lua
---------------------------------------------------------------------------
-- Wicked widgets for the awesome window manager
---------------------------------------------------------------------------
-- Lucas de Vries <lucas@glacicle.com>
-- Licensed under the WTFPL
-- Version: v1.0pre-awe3.0rc4
---------------------------------------------------------------------------

-- Require libs
require("awful")

---- {{{ Grab environment
local ipairs = ipairs
local pairs = pairs
local print = print
local type = type
local tonumber = tonumber
local tostring = tostring
local math = math
local table = table
local awful = awful
local os = os
local io = io
local string = string

-- Grab C API
local capi =
{
    awesome = awesome,
    screen = screen,
    client = client,
    mouse = mouse,
    button = button,
    titlebar = titlebar,
    widget = widget,
    hooks = hooks,
    keygrabber = keygrabber
}

-- }}}

-- Wicked: Widgets for the awesome window manager
module("wicked")

---- {{{ Initialise variables
local registered = {}
local widget_cache = {}

-- Initialise function tables
widgets = {}
helper = {}

local nets = {}
local cpu_total = {}
local cpu_active = {}
local cpu_usage = {}

-- }}}

---- {{{ Helper functions

----{{{ Max width
function helper.max_width(str, width)
    l = str:len()
    
    if l > width then
        r = math.floor(width/2)
        a = str:sub(1,r)
        b = str:sub(l-r, l)
        str = a .. "..." .. b
    end

    return str
end
----}}}

----{{{ Force a fixed width on a string with spaces
function helper.fixed_width(str, width)
    l = str:len()
    n = width-l
    if n >= 0 then
        for i = 1, n do
            str = str.." "
        end
    else
        str = str:sub(0, l+n)
    end
    return str
end
----}}}

---- {{{ Format a string with args
function helper.format(format, args)
    -- TODO: Find a more efficient way to do this

    -- Format a string
    for var,val in pairs(args) do
        format = string.gsub(format, '$'..var, val)
    end

    -- Return formatted string
    return format
end
-- }}}

---- {{{ Padd a number to a minimum amount of digits
function helper.padd(number, padding)
    s = tostring(number)

    if padding == nil then
        return s
    end

    for i=1,padding do
        if math.floor(number/math.pow(10,(i-1))) == 0 then
            s = "0"..s
        end
    end

    if number == 0 then
        s = s:sub(2)
    end

    return s
end
-- }}}

---- {{{ Convert amount of bytes to string
function helper.bytes_to_string(bytes, sec, padding)
    if bytes == nil or tonumber(bytes) == nil then
        return ''
    end

    bytes = tonumber(bytes)

    signs = {}
    signs[1] = '  b'
    signs[2] = 'KiB'
    signs[3] = 'MiB'
    signs[4] = 'GiB'
    signs[5] = 'TiB'

    sign = 1

    while bytes/1024 > 1 and signs[sign+1] ~= nil do
        bytes = bytes/1024
        sign = sign+1
    end

    bytes = bytes*10
    bytes = math.floor(bytes)/10

    if padding then
        bytes = helper.padd(bytes*10, padding+1)
        bytes = bytes:sub(1, bytes:len()-1).."."..bytes:sub(bytes:len())
    end

    if sec then
        return tostring(bytes)..signs[sign]..'ps'
    else
        return tostring(bytes)..signs[sign]
    end
end
-- }}}

---- {{{ Split by whitespace
function helper.splitbywhitespace(str)
    values = {}
    start = 1
    splitstart, splitend = string.find(str, ' ', start)
    
    while splitstart do
        m = string.sub(str, start, splitstart-1)
        if m:gsub(' ','') ~= '' then
            table.insert(values, m)
        end

        start = splitend+1
        splitstart, splitend = string.find(str, ' ', start)
    end

    m = string.sub(str, start)
    if m:gsub(' ','') ~= '' then
        table.insert(values, m)
    end

    return values
end
-- }}}

--{{{ Escape a string
function helper.escape(text)
    if text then
        text = text:gsub("&", "&amp;")
        text = text:gsub("<", "&lt;")
        text = text:gsub(">", "&gt;")
        text = text:gsub("'", "&apos;")
        text = text:gsub("\"", "&quot;")
    end
    return text
end

-- }}}

-- }}}

---- {{{ Widget types

---- {{{ MPD widget type
function widgets.mpd()
    ---- Get data from mpc
    local nowplaying_file = io.popen('mpc')
    local nowplaying = nowplaying_file:read()

    -- Check that it's not nil
    if nowplaying == nil then
        return {''}
    end

    -- Close the command
    nowplaying_file:close()
    
    -- Escape
    nowplaying = helper.escape(nowplaying)

    -- Return it
    return {nowplaying}
end

widget_cache[widgets.mpd] = {}
-- }}}

---- {{{ MOCP widget type
function widgets.mocp(format, max_width)
	local playing = ''

    ---- Get data from mocp
    local info = io.popen('mocp -i')
    local state = info:read()
	state = state.gsub(state, 'State: ', '')

	if (state == "PLAY") then
		local file = info:read()
		file = file.gsub(file, 'File: ', '')
		local title = info:read()
		title = title.gsub(title, 'Title: ', '')
		local artist = info:read()
		artist = artist.gsub(artist, 'Artist: ', '')
		local songtitle = info:read()
		songtitle = songtitle.gsub(songtitle, 'SongTitle: ', '')
		local album = info:read()
		album = album.gsub(album, 'Album: ', '')
		
		-- Try artist - (song)title
		if (artist:len() > 0) then
			playing = artist .. ' - ' .. (songtitle ~= '' and songtitle or title)
			
		-- Else try title or songtitle
		elseif (artist:len() == 0 and (title:len() > 0 or songtitle:len() > 0)) then
			playing = (title ~= '' and title or songtitle)

		-- Else use the filename
		else
			file = string.reverse(file)
			i = string.find(file, '/')
			if (i ~= nil) then
				file = string.sub(file, 0, i-1)
			end
			playing = string.reverse(file)
		end
	else
		playing = state
	end

	-- Close file
	info:close()

	-- Apply maximum width
	if (max_width ~= nil) then
		playing = helper.max_width(playing, max_width)
	end

	playing = helper.escape(playing)

    -- Return it
    return {playing}
end

widget_cache[widgets.mocp] = {}
-- }}}

---- {{{ CPU widget type
function widgets.cpu(format, padding)
    -- Calculate CPU usage for all available CPUs / cores and return the
    -- usage

    -- Perform a new measurement
    ---- Get /proc/stat
    local cpu_lines = {}
    local cpu_usage_file = io.open('/proc/stat')
    for line in cpu_usage_file:lines() do
        if string.sub(line, 1, 3) == 'cpu' then
            table.insert(cpu_lines, helper.splitbywhitespace(line))
        end
    end
    cpu_usage_file:close()

    ---- Ensure tables are initialized correctly
    while #cpu_total < #cpu_lines do
        table.insert(cpu_total, 0)
    end
    while #cpu_active < #cpu_lines do
        table.insert(cpu_active, 0)
    end
    while #cpu_usage < #cpu_lines do
        table.insert(cpu_usage, 0)
    end

    ---- Setup tables
    total_new     = {}
    active_new    = {}
    diff_total    = {}
    diff_active   = {}

    for i,v in ipairs(cpu_lines) do
        ---- Calculate totals
        total_new[i]    = 0
        for j = 2, #v do
            total_new[i] = total_new[i] + v[j]
        end
        active_new[i]   = v[2] + v[3] + v[4]
    
        ---- Calculate percentage
        diff_total[i]   = total_new[i]  - cpu_total[i]
        diff_active[i]  = active_new[i] - cpu_active[i]
        cpu_usage[i]    = math.floor(diff_active[i] / diff_total[i] * 100)

        ---- Store totals
        cpu_total[i]    = total_new[i]
        cpu_active[i]   = active_new[i]
    end

    if padding ~= nil then
        for k,v in pairs(cpu_usage) do
            if type(padding) == "table" then
                p = padding[k]
            else
                p = padding
            end

            cpu_usage[k] = helper.padd(cpu_usage[k], p)
        end
    end

    return cpu_usage
end

widget_cache[widgets.cpu] = {}
-- }}}

---- {{{ Memory widget type
function widgets.mem(format, padding)
    -- Return MEM usage values
    local f = io.open('/proc/meminfo')

    ---- Get data
    for line in f:lines() do
        line = helper.splitbywhitespace(line)

        if line[1] == 'MemTotal:' then
            mem_total = math.floor(line[2]/1024)
        elseif line[1] == 'MemFree:' then
            free = math.floor(line[2]/1024)
        elseif line[1] == 'Buffers:' then
            buffers = math.floor(line[2]/1024)
        elseif line[1] == 'Cached:' then
            cached = math.floor(line[2]/1024)
        end
    end
    f:close()

    ---- Calculate percentage
    mem_free=free+buffers+cached
    mem_inuse=mem_total-mem_free
    mem_usepercent = math.floor(mem_inuse/mem_total*100)

    if padding then
        if type(padding) == "table" then
            mem_usepercent = helper.padd(mem_usepercent, padding[1])
            mem_inuse = helper.padd(mem_inuse, padding[2])
            mem_total = helper.padd(mem_total, padding[3])
            mem_free = helper.padd(mem_free, padding[4])
        else
            mem_usepercent = helper.padd(mem_usepercent, padding)
            mem_inuse = helper.padd(mem_inuse, padding)
            mem_total = helper.padd(mem_total, padding)
            mem_free = helper.padd(mem_free, padding)
        end
    end

    return {mem_usepercent, mem_inuse, mem_total, mem_free}
end

widget_cache[widgets.mem] = {}
-- }}}

---- {{{ Swap widget type
function widgets.swap(format, padding)
    -- Return SWAP usage values
    local f = io.open('/proc/meminfo')

    ---- Get data
    for line in f:lines() do
        line = helper.splitbywhitespace(line)

        if line[1] == 'SwapTotal:' then
            swap_total = math.floor(line[2]/1024)
        elseif line[1] == 'SwapFree:' then
            free = math.floor(line[2]/1024)
        elseif line[1] == 'SwapCached:' then
            cached = math.floor(line[2]/1024)
        end
    end
    f:close()

    ---- Calculate percentage
    swap_free=free+cached
    swap_inuse=swap_total-swap_free
    swap_usepercent = math.floor(swap_inuse/swap_total*100)

    if padding then
        if type(padding) == "table" then
            swap_usepercent = helper.padd(swap_usepercent, padding[1])
            swap_inuse = helper.padd(swap_inuse, padding[2])
            swap_total = helper.padd(swap_total, padding[3])
            swap_free = helper.padd(swap_free, padding[4])
        else
            swap_usepercent = helper.padd(swap_usepercent, padding)
            swap_inuse = helper.padd(swap_inuse, padding)
            swap_total = helper.padd(swap_total, padding)
            swap_free = helper.padd(swap_free, padding)
        end
    end

    return {swap_usepercent, swap_inuse, swap_total, swap_free}
end

widget_cache[widgets.swap] = {}
-- }}}

---- {{{ Date widget type
function widgets.date(format)
    -- Get format
    if format == nil then
        return os.date()
    else
        return os.date(format)
    end
end
-- }}}

---- {{{ Filesystem widget type
function widgets.fs(format, padding)
    local f = io.popen('df -hP')
    local args = {}

    for line in f:lines() do
        vars = helper.splitbywhitespace(line)
        
        if vars[1] ~= 'Filesystem' and #vars == 6 then
            vars[5] = vars[5]:gsub('%%','')

            if padding then
                if type(padding) == "table" then
                    vars[2] = helper.padd(vars[2], padding[1])
                    vars[3] = helper.padd(vars[3], padding[2])
                    vars[4] = helper.padd(vars[4], padding[3])
                    vars[5] = helper.padd(vars[5], padding[4])
                else
                    vars[2] = helper.padd(vars[2], padding)
                    vars[3] = helper.padd(vars[3], padding)
                    vars[4] = helper.padd(vars[4], padding)
                    vars[5] = helper.padd(vars[5], padding)
                end
            end

            args['{'..vars[6]..' size}'] = vars[2]
            args['{'..vars[6]..' used}'] = vars[3]
            args['{'..vars[6]..' avail}'] = vars[4]
            args['{'..vars[6]..' usep}'] = vars[5]
        end
    end

    f:close()
    return args
end
-- }}}

---- {{{ Net widget type
function widgets.net(format, padding)
    local f = io.open('/proc/net/dev')
    args = {}

    for line in f:lines() do
        line = helper.splitbywhitespace(line)

        local p = line[1]:find(':')
        if p ~= nil then
            name = line[1]:sub(0,p-1)
            line[1] = line[1]:sub(p+1)

            if tonumber(line[1]) == nil then
                line[1] = line[2]
                line[9] = line[10]
            end

            if padding then
                args['{'..name..' rx}'] = helper.bytes_to_string(line[1], nil, padding)
                args['{'..name..' tx}'] = helper.bytes_to_string(line[9], nil, padding)
            else
                args['{'..name..' rx}'] = helper.bytes_to_string(line[1])
                args['{'..name..' tx}'] = helper.bytes_to_string(line[9])
            end

            args['{'..name..' rx_b}'] = math.floor(line[1]*10)/10
            args['{'..name..' tx_b}'] = math.floor(line[9]*10)/10
            
            args['{'..name..' rx_kb}'] = math.floor(line[1]/1024*10)/10
            args['{'..name..' tx_kb}'] = math.floor(line[9]/1024*10)/10

            args['{'..name..' rx_mb}'] = math.floor(line[1]/1024/1024*10)/10
            args['{'..name..' tx_mb}'] = math.floor(line[9]/1024/1024*10)/10

            args['{'..name..' rx_gb}'] = math.floor(line[1]/1024/1024/1024*10)/10
            args['{'..name..' tx_gb}'] = math.floor(line[9]/1024/1024/1024*10)/10

            if nets[name] == nil then 
                nets[name] = {}
                args['{'..name..' down}'] = 'n/a'
                args['{'..name..' up}'] = 'n/a'
                
                args['{'..name..' down_b}'] = 0
                args['{'..name..' up_b}'] = 0

                args['{'..name..' down_kb}'] = 0
                args['{'..name..' up_kb}'] = 0

                args['{'..name..' down_mb}'] = 0
                args['{'..name..' up_mb}'] = 0

                args['{'..name..' down_gb}'] = 0
                args['{'..name..' up_gb}'] = 0

                nets[name].time = os.time()
            else
                interval = os.time()-nets[name].time
                nets[name].time = os.time()

                down = (line[1]-nets[name][1])/interval
                up = (line[9]-nets[name][2])/interval

                if padding then
                    args['{'..name..' down}'] = helper.bytes_to_string(down, true, padding)
                    args['{'..name..' up}'] = helper.bytes_to_string(up, true, padding)
                else
                    args['{'..name..' down}'] = helper.bytes_to_string(down, true)
                    args['{'..name..' up}'] = helper.bytes_to_string(up, true)
                end

                args['{'..name..' down_b}'] = math.floor(down*10)/10
                args['{'..name..' up_b}'] = math.floor(up*10)/10

                args['{'..name..' down_kb}'] = math.floor(down/1024*10)/10
                args['{'..name..' up_kb}'] = math.floor(up/1024*10)/10

                args['{'..name..' down_mb}'] = math.floor(down/1024/1024*10)/10
                args['{'..name..' up_mb}'] = math.floor(up/1024/1024*10)/10

                args['{'..name..' down_gb}'] = math.floor(down/1024/1024/1024*10)/10
                args['{'..name..' up_gb}'] = math.floor(up/1024/1024/1024*10)/10
            end

            nets[name][1] = line[1]
            nets[name][2] = line[9]
        end
    end

    f:close()
    return args
end
widget_cache[widgets.net] = {}
-- }}}

---- {{{ Uptime widget type
function widgets.uptime(format, padding)
    --Get uptime from /proc/uptime
	local f = io.open("/proc/uptime")
	uptime_line = f:read()
	
	f:close()

	args = {}
    --/proc/uptime has the format "<up time> <idle time>"
  	if uptime_line:find(" ") ~= nil then

		pend = uptime_line:find(" ",0,true)
                
        	uptime_line_part = uptime_line:sub(0,pend-1)		

		total_uptime  = math.floor( tonumber(uptime_line_part) )

		uptime_days    =  math.floor( total_uptime / (3600 * 24) )
		uptime_hours   =  math.floor( ( total_uptime % (3600 * 24) ) / 3600 )
       	 	uptime_minutes =  math.floor( ( ( total_uptime % (3600 * 24) ) % 3600 ) / 60 )
        	uptime_seconds =  math.floor( ( ( total_uptime % (3600  * 24) ) % 3600) % 60 )

		if padding then
			
			if type(padding) == "table" then
				total_uptime   = helper.padd(total_uptime   , padding[1])
				uptime_days    = helper.padd(uptime_days    , padding[2])
				uptime_hours   = helper.padd(uptime_hours   , padding[3])
 				uptime_minutes = helper.padd(uptime_minutes , padding[4])
				uptime_seconds = helper.padd(uptime_seconds , padding[5])
			else
				total_uptime   = helper.padd(total_uptime   , padding)
				uptime_days    = helper.padd(uptime_days    , padding)
				uptime_hours   = helper.padd(uptime_hours   , padding)
				uptime_minutes = helper.padd(uptime_minutes , padding)
  				uptime_seconds = helper.padd(uptime_seconds , padding)
			end	
		
		end
	end
	
	return {total_uptime, uptime_days, uptime_hours, uptime_minutes, uptime_seconds}

end
widget_cache[widgets.uptime] = {}
-- }}}

-- For backwards compatibility: custom function
widgets["function"] = function ()
    return {}
end

-- }}}

---- {{{ Main functions
---- {{{ Register widget
function register(widget, wtype, format, timer, field, padd)
    local reg = {}
    local widget = widget

    -- Set properties
    reg.type = wtype
    reg.format = format
    reg.timer = timer
    reg.field = field
    reg.padd = padd
    reg.widget = widget

    -- Update function
    reg.update = function ()
        update(widget, reg)
    end

    -- Default to timer=1
    if reg.timer == nil then
        reg.timer = 1
    end

    -- Allow using a string widget type
    if type(reg.type) == "string" then
        reg.type = widgets[reg.type]
    end

    -- Register reg object
    regregister(reg)

    -- Return reg object for reuse
    return reg
end
-- }}}

-- {{{ Register from reg object
function regregister(reg)
    if not reg.running then
        -- Put widget in table
        if registered[reg.widget] == nil then
            registered[reg.widget] = {}
            table.insert(registered[reg.widget], reg)
        else
            already = false

            for w, i in pairs(registered) do
                if w == reg.widget then
                    for k,v in pairs(i) do
                        if v == reg then
                            already = true
                            break
                        end
                    end

                    if already then
                        break
                    end
                end
            end

            if not already then
                table.insert(registered[reg.widget], reg)
            end
        end

        -- Start timer
        awful.hooks.timer.register(reg.timer, reg.update)

        -- Initial update
        reg.update()

        -- Set running
        reg.running = true
    end
end
-- }}}

-- {{{ Unregister widget
function unregister(widget, keep, reg)
    if reg == nil then
        for w, i in pairs(registered) do
            if w == widget then
                for k,v in pairs(i) do
                    reg = unregister(w, keep, v)
                end
            end
        end

        return reg
    end

    if not keep then
        for w, i in pairs(registered) do
            if w == widget then
                for k,v in pairs(i) do
                    if v == reg then
                        table.remove(registered[w], k)
                    end
                end
            end
        end
    end

    awful.hooks.timer.unregister(reg.update)

    reg.running = false
    return reg
end
-- }}}

-- {{{ Suspend wicked, halt all widget updates
function suspend()
    for w, i in pairs(registered) do
        for k,v in pairs(i) do
            unregister(w, true, v)
        end
    end
end
-- }}}

-- {{{ Activate wicked, restart all widget updates
function activate(widget)
    for w, i in pairs(registered) do
        if widget == nil or w == widget then
            for k,v in pairs(i) do
                regregister(v)
            end
        end
    end
end
-- }}}

-- {{{ Enable caching for a widget type
function enable_caching(widget)
    if widget_cache[widget] == nil then
        widget_cache[widget] = {}
    end
end
-- }}}

---- {{{ Update widget
function update(widget, reg)
    -- Check if there are any equal widgets
    if reg == nil then
        for w, i in pairs(registered) do
            if w == widget then
                for k,v in pairs(i) do
                    update(w, v)
                end
            end
        end

        return
    end

    local t = os.time()
    local data = {}

    -- Check if we have output chached for this widget,
    -- newer than last widget update.
    if widget_cache[reg.type] ~= nil then
        local c = widget_cache[reg.type]

        if c.time == nil or c.time <= t-reg.timer then
            c.time = t
            c.data = reg.type(reg.format, reg.padd)
        end
        
        data = c.data
    else
        data = reg.type(reg.format, reg.padd)
    end

    if type(data) == "table" then
        if type(reg.format) == "string" then
            data = helper.format(reg.format, data)
        elseif type(reg.format) == "function" then
            data = reg.format(widget, data)
        end
    end
    
    if reg.field == nil then
        widget.text = data
    elseif widget.plot_data_add ~= nil then
        widget:plot_data_add(reg.field, tonumber(data))
    elseif widget.bar_data_add ~= nil then
        widget:bar_data_add(reg.field, tonumber(data))
    end
    return data
end

-- }}}

-- }}}

-- vim: set filetype=lua fdm=marker tabstop=4 shiftwidth=4 nu:
